mitiq/executor/executor.py

```diff
@@ -160,38 +185,51 @@ def evaluate(
                 for circuit_with_measurements in observable.measure_in(circuit)
             ]
             result_step = observable.ngroups
-        elif (
-            observable is not None
-            and self._executor_return_type not in MeasurementResultLike
-            and self._executor_return_type not in DensityMatrixLike
-        ):
-            raise ValueError(
-                """Executor and observable are not compatible. Executors
-                returning expectation values as float must be used with
-                observable=None"""
-            )
         else:
             all_circuits = circuits
             result_step = 1
 
         # Run all required circuits.
-        all_results = self.run(all_circuits, force_run_all, **kwargs)
+        try:
+            all_results = self.run(all_circuits, force_run_all, **kwargs)
+        except Exception:
+            if observable is not None and self._executor_return_type is None:
+                all_circuits = [
+                    circuit_with_measurements
+                    for circuit in circuits
+                    for circuit_with_measurements in observable.measure_in(
+                        circuit
+                    )
+                ]
+                all_results = self.run(all_circuits, force_run_all, **kwargs)
+            else:
+                raise
```

Comment:
I like the idea of just executing one of the circuits and checking the return type. Does this mean we can move all the compatibility checks in lines 156-175 after this manual checking of return type? This would remove the need of _executor_return_type variable and no typehinting is needed anymore. Maybe I am missing something.

Reply:
A couple of things.
1. This is NOT checking the return type of a single circuit, but instead this is running all circuits with the `executor`. 

2. The checks in lines 156-175 are specifically checking the following cases based off limitations already in the `execute` function.

- FloatLike - expected the circuits to already include measurements and there should NOT be an observable. The `executor` does not handle observables for FloatLike results.
- DensityMatrixLike and MeasurementLike - MeasurementResultLike handles observables in lines 182-187 and both respectively use OBSERVABLE.`_expectation_from_density_matrix()` and OBSERVABLE.`_expectation_from_measurements()`

Since these are requirements, I would think it would be better to check and break out in bad scenarios, instead of trying to run the circuits and then error out.

Reply:
Ah yes, you are right that it is running all circuits! I guess I was looking at lines 209-216 where you use the manual return type (of the first circuit?) to parse results. I was thinking to just run the first circuit to get the manual return type and do all the checks on observable compatibility afterwards. This is a completely different approach than using _executor_return_type, so understandable if we don't want to that.

Reply:
The main use of `_executor_return_type` (outside of what I have done in this PR) is to check if it is `MeasurementResultLike` and there is an observable so that `observable.measure_in()` can be called to add measurements into the circuit. Without doing this, the backends complain. I can't think of a way to do this without the `_executor_return_type` since calling `run()` to try to get back a manual return type throws errors.